lib/src/internal/http/http_response.dart

```diff
@@ -1,107 +1,168 @@
 import 'dart:convert';
+import 'dart:typed_data';
 
 import 'package:http/http.dart' as http;
+import 'package:nyxx/nyxx.dart';
+import 'package:nyxx/src/internal/response_wrapper/error_response_wrapper.dart';
 
+/// Represents a HTTP result from the API.
 abstract class IHttpResponse {
+  /// The status code of the response.
   int get statusCode;
+
+  /// The headers associated with the response.
   Map<String, String> get headers;
+
+  /// The body of this response.
+  ///
+  /// See also:
+  /// - [textBody], for getting the body decoded as a [String].
+  /// - [jsonBody], for decoding the body decoded as JSON.
+  Uint8List get body;
+
+  /// The body of this response, decoded as UTF-8.
+  ///
+  /// Will be `null` if the body was not valid UTF-8.
+  ///
+  /// See also:
+  /// - [jsonBody], for getting the body decoded as JSON.
+  String? get textBody;
+
+  /// The body of this response, decoded as JSON.
+  ///
+  /// Will be `null` is the body was not valid JSON, and [hasJsonBody] will be set to `false`.
+  dynamic get jsonBody;
+
+  /// Whether this response has a JSON body.
+  bool get hasJsonBody;
 }
 
 abstract class HttpResponse implements IHttpResponse {
   @override
-  late final int statusCode;
+  final int statusCode;
+
   @override
-  late final Map<String, String> headers;
+  final Map<String, String> headers;
 
-  late final List<int> _body;
-  late final dynamic _jsonBody;
-  late final http.ByteStream _bodyStream;
+  @override
+  final Uint8List body;
 
-  /// Creates an instance of [HttpResponse]
-  HttpResponse(http.StreamedResponse response) {
-    _bodyStream = response.stream;
-    statusCode = response.statusCode;
-    headers = response.headers;
-  }
+  @override
+  late final String? textBody;
+
+  @override
+  late final dynamic jsonBody;
+
+  @override
+  late final bool hasJsonBody;
+
+  final http.BaseResponse response;
+  http.BaseRequest get request => response.request!;
 
-  Future<void> finalize() async {
-    _body = await _bodyStream.toBytes();
+  HttpResponse({
+    required this.response,
+    required this.body,
+  })  : statusCode = response.statusCode,
+        headers = response.headers {
+    String? textBody;
+    dynamic jsonBody;
+    bool hasJsonBody = false;
 
     try {
-      if (_body.isNotEmpty) {
-        _jsonBody = jsonDecode(utf8.decode(_body));
-      } else {
-        _jsonBody = null;
-      }
+      textBody = utf8.decode(body);
+      jsonBody = jsonDecode(textBody);
```

Comment:
here we are storing "same" data twice. It is realistic that we are gonna access json data more than once? If not we could do getter and decode on the fly to presever some memory

Reply:
No, but because HttpResponse objects are short lived anyways, we don't gain any memory benefit from storing it for a short amount of time against decoding on the fly.

In fact, HttpResponseError does access jsonBody more than once, so we save on the CPU cost of decoding it twice by storing it.